romdisk/uart.asm

```diff
@@ -124,23 +128,66 @@ uartsnd_main:
         ld l, a
         ; HL contains the string to output on the UART
         ; Get its length and save it in BC
-        push hl
-        call strlen
-        push bc
+
+        ; Filename in HL
+        ld b, h
+        ld c, l
+        ld h, O_RDONLY
+        OPEN()
+        or a
+        jp m, uart_file_error
+        ld d, a
+
         call uart_open
-        jp m, uart_error_pop
-        ; Write to the UART
-        ; DE - Buffer
-        ; BC - Size
-        ; H  - Descriptor
-        pop bc
+        ld e, a
+
+        push de
+_uartsnd_loop:
         pop de
-        ld h, a
-        WRITE()
-        ; Close the driver
+        push de
+        ; Read from source file and write to destination file, until we don't have any bytes
+        S_READ3(d, STATIC_BUFFER, STATIC_BUFFER_SIZE)
+        or a
+        jr nz, _uart_error_close_all
+        ; Check if we there are no bytes to read anymore
+        ld a, b
+        or c
+        jr z, _uartsnd_end
+        ; Get the destination descriptor back but keep it saved
+        pop de
+        push de
+        ; BC already contains the number of bytes to write
+        S_WRITE2(e, STATIC_BUFFER)
+        ; In theory we should check that we wrote the same amount of bytes, but for
+        ; simplicity reasons, let's only check for errors
+        or a
+        jr nz, _uart_error_close_all
+        jp _uartsnd_loop
+_uartsnd_end:
+        ; Close the opened descriptors
+        pop de
+        ld h, d
+        CLOSE()
+        ld h, e
         CLOSE()
+        xor a   ; Success
         ret
 
+_uart_error_close_all:
+        pop de
+        ld b, a
+        ld h, e
+        CLOSE()
```

Comment:
You are not closing the UART driver here, only the file. Since it would be very similar to the function above, try something like:
```
_uartsnd_end:
        pop de
_uartsnd_close:
        ; Close the opened descriptors
        ld h, d
        CLOSE()
        ld h, e
        CLOSE()
        xor a   ; Success
        ret

_uart_error_close_all:
        pop de
        ld b, a
        call _uartsnd_close
        ld a, b
        jp error_print
```

Reply:
#replaced